internal/resources/iampolicy/scope_helper.go

```diff
@@ -0,0 +1,156 @@
+	return [...]string{
+		"unknown",
+		organizationKey,
+		clusterGroupKey,
+		clusterKey,
+		workspaceKey,
+		namespaceKey,
+	}[s]
+}
+
+func validateScope(ctx context.Context, diff *schema.ResourceDiff, i interface{}) error {
+	value, ok := diff.GetOk(scopeKey)
+	if !ok {
+		return fmt.Errorf("scope: %v is not valid: minimum one valid scope block is required", value)
+	}
+
+	data, _ := value.([]interface{})
+
+	if len(data) == 0 || data[0] == nil {
+		return fmt.Errorf("scope data: %v is not valid: minimum one valid scope block is required", data)
+	}
+
+	scopeData := data[0].(map[string]interface{})
+	scopesFound := make([]string, 0)
+	scopesAllowed := []string{organizationKey, clusterGroupKey, clusterKey, workspaceKey, namespaceKey}
+
+	if v, ok := scopeData[organizationKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, organizationKey)
+		}
+	}
+
+	if v, ok := scopeData[clusterGroupKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, clusterGroupKey)
+		}
+	}
+
+	if v, ok := scopeData[clusterKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, clusterKey)
+		}
+	}
+
+	if v, ok := scopeData[workspaceKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, workspaceKey)
+		}
+	}
+
+	if v, ok := scopeData[namespaceKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopesFound = append(scopesFound, namespaceKey)
+		}
+	}
+
+	if len(scopesFound) == 0 {
+		return fmt.Errorf("no valid scope type block found: minimum one valid scope type block is required among: %v", scopesAllowed)
+	} else if len(scopesFound) > 1 {
+		return fmt.Errorf("found scopes: %v are not valid: maximum one valid scope type block is allowed", scopesFound)
+	}
+
+	return nil
+}
+
+func getScopeType(d *schema.ResourceData) (scopeType, error) {
+	scopes := []scopeType{}
+
+	value, ok := d.GetOk(scopeKey)
+	if !ok {
+		return unknown, fmt.Errorf("no scope defined")
+	}
+
+	data, _ := value.([]interface{})
+
+	if len(data) == 0 || data[0] == nil {
+		return unknown, fmt.Errorf("no type defined in scope")
+	}
+
+	scopeData, _ := data[0].(map[string]interface{})
+
+	for key, val := range scopeData {
+		v := val.([]interface{})
+		if len(v) == 0 {
+			continue
+		}
+
+		scopes = append(scopes, stringToScope(key))
+	}
+
+	if len(scopes) != 1 {
+		return unknown, fmt.Errorf("none or more than one scope types are defined")
+	}
+
+	return scopes[0], nil
+}
+
+func getIntersectionOfSubs(
+	state, server []*iammodel.VmwareTanzuCoreV1alpha1PolicySubject,
+) []*iammodel.VmwareTanzuCoreV1alpha1PolicySubject {
+	newList := []*iammodel.VmwareTanzuCoreV1alpha1PolicySubject{}
+	for _, each := range state {
+		for _, sub := range server {
+			if each.Name == sub.Name && *each.Kind == *sub.Kind {
+				newList = append(newList, each)
+			}
+		}
+	}
+
+	return newList
+}
```

Comment:
Consider using maps for finding intersections.

Also, this function may not be a part of the file `scope_helper.go`. Consider renaming it to splitting this file into two.